dash/_hooks.py

```diff
@@ -0,0 +1,93 @@
+import typing as _t
+
+_ns = {
+    "setup": [],
+    "layout": [],
+    "routes": [],
+    "error": [],
+    "callback": [],
+}
+
+
+def layout(func):
+    """
+    Run a function when serving the layout, the return value
+    will be used as the layout.
+    """
+    _ns["layout"].append(func)
+    return func
+
+
+def setup(func):
+    """
+    Can be used to get a reference to the app after it is instantiated.
+    """
+    _ns["setup"].append(func)
+    return func
+
+
+def route(name=None, methods=("GET",)):
+    """
+    Add a route to the Dash server.
+    """
+
+    def wrap(func):
+        _name = name or func.__name__
+        _ns["routes"].append((_name, func, methods))
```

Comment:
hm - so some of the `_ns` entries are functions and some are tuples, so the structure of entries in `_ns["key"]` depends on the value of `"key"`?

Reply:
The `_ns` just saves the info for the different hooks, it isn't meant for public usage. For routes/callbacks it needs more arguments so it's saved in a tuple for easy unpacking.